slsim/Sources/QuasarCatalog/quasar_pop.py

```diff
@@ -0,0 +1,355 @@
+            redshift=self.redshifts,
+            density=n_comoving_values,
+            sky_area=self.sky_area,
+            cosmology=self.cosmo,
+            noise=self.noise,
+        )
+
+        # Ensure redshifts are stored as numpy array
+        sampled_redshifts = np.array(sampled_redshifts, dtype=float)
+
+        if return_single:
+            return np.random.choice(sampled_redshifts)
+
+        return sampled_redshifts
+
+    def compute_cdf_data(self, m_min, m_max, quasar_redshifts):
+        """Computes cumulative distribution function (CDF) data for given redshift
+        values.
+
+        :param m_min: Minimum apparent magnitude.
+        :type m_min: float
+        :param m_max: Maximum apparent magnitude.
+        :type m_max: float
+        :param quasar_redshifts: Redshift values generated from `generate_quasar_redshifts`.
+        :type quasar_redshifts: array-like
+        :return: Dictionary containing CDF data for each redshift.
+        :rtype: dict
+        """
+        cdf_data_dict = {}
+
+        for z in np.unique(quasar_redshifts):
+            M_min = self.convert_magnitude(m_min, z, conversion="apparent_to_absolute")
+            M_max = self.convert_magnitude(m_max, z, conversion="apparent_to_absolute")
+
+            M_values = np.linspace(M_min, M_max, 100)
+
+            dPhi_dM_values = np.array([self.dPhi_dM(M, z) for M in M_values])
+
+            sorted_M_values = np.sort(M_values)
+            cumulative_probabilities = np.cumsum(dPhi_dM_values)
+            max_cumulative_probabilities = np.max(cumulative_probabilities)
+            cumulative_prob_norm = (
+                cumulative_probabilities / max_cumulative_probabilities
+            )
+
+            cdf_data_dict[z] = (sorted_M_values, cumulative_prob_norm)
+
+        return cdf_data_dict
+
+    def inverse_cdf_fits_for_redshifts(self, m_min, m_max, quasar_redshifts):
+        """Creates inverse Cumulative Distribution Function (CDF) fits for each
+        redshift.
+
+        :param m_min: Minimum apparent magnitude.
+        :type m_min: float
+        :param m_max: Maximum apparent magnitude.
+        :type m_max: float
+        :param quasar_redshifts: Redshift values generated from `generate_quasar_redshifts`.
+        :type quasar_redshifts: array-like
+        :return: Dictionary containing inverse CDF functions for each redshift.
+        :rtype: dict
+        """
+        cdf_data = self.compute_cdf_data(m_min, m_max, quasar_redshifts)
+        inverse_cdf_dict = {}
+
+        for redshift, (sorted_M_values, cumulative_prob_norm) in cdf_data.items():
+            inverse_cdf = interp1d(
+                cumulative_prob_norm,
+                sorted_M_values,
+                kind="linear",
+                fill_value="extrapolate",
+            )
+            inverse_cdf_dict[redshift] = inverse_cdf
+
+        return inverse_cdf_dict
+
+    def quasar_sample(self, m_min, m_max, quasar_redshifts, seed=42):
```

Comment:
I think you should remove quasar_redshifts from the input. Please,  put quasar_redshifts = self.generate_quasar_redshifts(m_min=m_min, m_max=m_max)) inside the function so that we do not need to provide redshift while calling this function. It will generate redshift sample and use it automatically.